Add check-for-reproducer action

kacperkapusciak opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

This PR adds a "check-for-reproducer" GitHub action proposed in https://github.com/facebook/react-native/issues/35591.
This GitHub action automatically labels an issue with when no link to GitHub repository under the authors name or link to Snack is present either in the issue body or in the comments.

## Changelog:

<!-- Help reviewers and the release process by writing your own changelog entry.

Pick one each for the category and type tags:

[ANDROID|GENERAL|IOS|INTERNAL] [BREAKING|ADDED|CHANGED|DEPRECATED|REMOVED|FIXED|SECURITY] - Message

For more details, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->

[INTERNAL] [ADDED] - Add check-for-reproducer GitHub action

## Test Plan:

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

This action was tested on a private repro mimicking on how GitHub actions are set-up in `facebook/react-native` repository. If you'd want to play around with the action on this private repo just ask for access - provided you're a maintainer. 

### Regarding issue body

1. Labels "Needs: Repro" when no link to Expo Snack or a Github repo under the author's name is present in the issue body:

![image](https://github.com/facebook/react-native/assets/39658211/b8e6d4aa-bc9a-471d-99b8-3b551f2c46c6)

2. Removes the "Needs: Repro" label and deletes the Missing Repro comment when the author edits the issue and provides a link to GitHub repo under their name

![image](https://github.com/facebook/react-native/assets/39658211/2dc13fae-b6ca-432f-88a2-f5c340f28c86)

3. Removes the "Needs: Repro" label and deletes the Missing Repro comment when the author edits the issue and provides a valid link to Expo Snack

![image](https://github.com/facebook/react-native/assets/39658211/e82e744e-b2fb-4979-84de-08f95686feb3)

### Regarding comments

3. Removes the "Needs: Repro" label and deletes the Missing Repro comment when there's a comment with a link to reproduction under issue author's name 

![image](https://github.com/facebook/react-native/assets/39658211/eefb5978-29d3-4486-8bd7-878fb2c07c05)

4. Removes the "Needs: Repro" label and deletes the Missing Repro comment when there's a comment with a link to Expo Snack

![image](https://github.com/facebook/react-native/assets/39658211/5769ff77-0d9f-4dac-bb36-6d257e6c2a35)

### Regarding false-positives

5. Adds a "Needs: Repro" label when a link to reprository isn't under the issue author's name

![image](https://github.com/facebook/react-native/assets/39658211/01c93ea3-88b9-4920-a8f8-bcd123d214de)

6. Adds a "Needs: Repro" label when a link to Expo Snack's homepage was sent

![image](https://github.com/facebook/react-native/assets/39658211/568c9824-67af-402a-b102-c1c355651d21)

7. Adds a "Needs: Repro" label when a random link was sent

![image](https://github.com/facebook/react-native/assets/39658211/23bb58b3-cc07-4284-a562-6651f2077a3c)

cortinico reviewed (approved):
Love it 🚀 Don't you mind editing also the "Needs: Repro" response message mentioning that we could not detect a repro in the submitted issue?

kacperkapusciak commented:
@cortinico I've updated the message 👍

NickGerleman commented on .github/workflow-scripts/checkForReproducer.js:
```diff
@@ -0,0 +1,79 @@
+    );
+    const hasGithubRepoLink = containsPattern(
+      body,
+      `https?:\\/\\/github\\.com\\/(${user})\\/[^/]+\\/?\\s?`,
```
nit: This might be a bit too constrained. I.e what if a user posts a repro for a project that belongs to an org, or an existing repro from another user?

kacperkapusciak replied:
It's a valid concern but my experience not that common. Implementing the bot so it would work when _any_ repo is present in the issue body would trigger too many false-positives.

I've seen more times that users tend to be "smart" about the repro and send the link to our repository like so:

![image](https://github.com/facebook/react-native/assets/39658211/a6de11f8-4524-499f-8063-75c86612720f)

Solid logic I guess - obviously the bug with Reanimated is in the Reanimated repo 🙄

Or even worse, have random repositories in the issue description:

![image](https://github.com/facebook/react-native/assets/39658211/e5574e30-4add-43e2-97f1-c270456d48d6)



Sure, I can implement as you suggest, but I'm pretty sure you won't be very happy about it long term

cortinico commented:
@kacperkapusciak This PR sadly broke a couple of things:

1. We now add this label also to PR, which we shouldn't (see https://github.com/facebook/react-native/pull/38522)
2. The label autoresponder doens't work anymore for `Needs: Repro`. (See: https://github.com/facebook/react-native/issues/38520)

Can you fix forward, otherwise we'll have to revert this to don't create unnecessary noise

kacperkapusciak commented:
@cortinico on it

kacperkapusciak commented:
@cortinico submitted a fix and an explaination to the first problem. It works inversely to [check in `autorebase` action](https://github.com/facebook/react-native/blob/main/.github/workflows/autorebase.yml#L17)

I'm still trying to reproduce the second problem in my sandbox env

kacperkapusciak commented:
@cortinico I think I have a solution and explanation of what went wrong for both problems but I need a bit more time to fully test the second PR (https://github.com/facebook/react-native/pull/38531)

NickGerleman commented:
> @kacperkapusciak This PR sadly broke a couple of things:
> 
> 1. We now add this label also to PR, which we shouldn't (see [cleanup RCTDisableTurboModuleManagerDelegateLocking #38522](https://github.com/facebook/react-native/pull/38522))
> 2. The label autoresponder doens't work anymore for `Needs: Repro`. (See: [After upgrading to 0.71.11 Xcode build fails with 'butter/map.h' file not found #38520](https://github.com/facebook/react-native/issues/38520))
> 
> Can you fix forward, otherwise we'll have to revert this to don't create unnecessary noise

#2 also has a repro, but not under the same name as the user.

I also realized, this restriction conflicts with the templates wording of “ Snack, code example, screenshot, or link to a repository”. So somebody may add a code sample as we requested and rudely get their issue closed.

kacperkapusciak commented:
@NickGerleman I can try to associate the autor of the comment with the owner of the repo. So, when someone comments but the repo is under their name it would be accepted as a reproduction. I think this is the way 👍 

On the other hand code samples are complicated - sometimes they are low quality and don't reproduce the issue. Eg. [here](https://github.com/facebook/react-native/issues/38530) Nicola decided to add `Needs: Repro` even with code sample present

Would you like me to try to detect JS code as a valid repro either way?

cortinico commented:
> I also realized, this restriction conflicts with the templates wording of “ Snack, code example, screenshot, or link to a repository”. So somebody may add a code sample as we requested and rudely get their issue closed.

@NickGerleman we should update the issue template and mention that it should be either a Snack or a repo, not a snippet (from a snippet to a Snack is literally a 1 minute operation).

cortinico commented:
Also @kacperkapusciak we should let the bot apply the "Needs: Repro" label only if "Needs: Triage" is there (or anyway not on closed issues: https://github.com/facebook/react-native/issues/35766#issuecomment-1644661365)

kacperkapusciak commented:
@cortinico I've added a code for preventing responding on closed issues in https://github.com/facebook/react-native/pull/38525 but the change didn't land on `main` (phabricator didn't pull the changes?). Moved the change to https://github.com/facebook/react-native/pull/38531 instead.

cortinico commented:
> but the change didn't land on `main` (phabricator didn't pull the changes?). Moved the change to #38531 instead.

Yup my bad I haven't noticed you pushed a new commit so I just merged the original one. I'll import the next one
